File: hoop_detector.py
import cv2
import numpy as np
from circle_detector import CircleDetector
import time

class HoopDetector:
    def __init__(self, kernel_size=5, solidity_threshold=0.6, min_score=0.7):
        """初始化篮筐检测器
        
        Args:
            kernel_size: 形态学操作的核大小
            solidity_threshold: 密实度阈值，用于区分篮筐和方块
            min_score: 圆检测的最小分数阈值
        """
        self.kernel_size = kernel_size
        self.solidity_threshold = solidity_threshold
        self.min_score = min_score
        self.kernel = np.ones((kernel_size, kernel_size), np.uint8)
        
        # 创建CircleDetector实例
        self.circle_detector = CircleDetector(
            num_rois=6,
            min_score=min_score
        )
        
        # 存储处理过程中的图像
        self.original_image = None
        self.initial_binary = None
        self.binary_image = None
        self.eroded_image = None
        self.contour_image = None
        self.filtered_image = None
        self.final_binary = None
        self.result_image = None
        
        # 存储各步骤的运行时间
        self.timing = {}
        
    def reset_timing(self):
        """重置时间统计"""
        self.timing = {}
        
    def load_image(self, image_path):
        """加载图像"""
        start_time = time.time()
        self.original_image = cv2.imread(image_path)
        if self.original_image is None:
            raise ValueError(f"无法加载图像: {image_path}")
        self.circle_detector.load_image(image_path)
        self.timing['load_image'] = time.time() - start_time
        return self
        
    def create_binary_image(self):
        """创建二值图像"""
        if self.original_image is None:
            raise ValueError("未加载图像")
        
        start_time = time.time()
        
        # 转换到HSV颜色空间并增强饱和度
        hsv_frame = cv2.cvtColor(self.original_image, cv2.COLOR_BGR2HSV)
        h, s, v = cv2.split(hsv_frame)
        s = np.clip(s * 1.5, 0, 255).astype(np.uint8)
        hsv_enhanced = cv2.merge([h, s, v])
        
        # 创建橙色掩码（初始二值图）
        lower_bound = np.array([5, 220, 0])
        upper_bound = np.array([11, 255, 255])
        self.initial_binary = cv2.inRange(hsv_enhanced, lower_bound, upper_bound)
        
        # 清理掩码
        kernel = np.ones((5, 5), np.uint8)
        # 开运算（先腐蚀后膨胀），去除小的噪点
        mask = cv2.morphologyEx(self.initial_binary, cv2.MORPH_OPEN, kernel)
        # 闭运算（先膨胀后腐蚀），填充小孔
        mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel)
        # 中值滤波，去除椒盐噪声
        self.binary_image = cv2.medianBlur(mask, 23)
        
        self.timing['create_binary'] = time.time() - start_time
        return self
        
    def process_image(self):
        """处理图像，分离篮筐和干扰物"""
        if self.binary_image is None:
            raise ValueError("未创建二值图像")
        
        start_time = time.time()
        
        # 1. 腐蚀操作，断开连接
        self.eroded_image = cv2.erode(self.binary_image, self.kernel, iterations=3)
        
        # 2. 查找轮廓
        contours, _ = cv2.findContours(self.eroded_image, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        
        # 创建轮廓图像（用于可视化）
        self.contour_image = cv2.cvtColor(self.eroded_image, cv2.COLOR_GRAY2BGR)
        cv2.drawContours(self.contour_image, contours, -1, (0, 255, 0), 2)
        
        # 3. 基于密实度筛选轮廓
        filtered_contours = []
        for contour in contours:
            # 计算轮廓面积和凸包
            area = cv2.contourArea(contour)
            hull = cv2.convexHull(contour)
            hull_area = cv2.contourArea(hull)
            
            # 计算密实度
            if hull_area > 0:
                solidity = float(area) / hull_area
                if solidity < self.solidity_threshold:
                    filtered_contours.append(contour)
        
        # 4. 创建过滤后的掩码
        self.filtered_image = np.zeros_like(self.binary_image)
        cv2.drawContours(self.filtered_image, filtered_contours, -1, 255, -1)
        
        # 5. 膨胀操作，恢复目标大小
        self.final_binary = cv2.dilate(self.filtered_image, self.kernel, iterations=2)
        
        self.timing['process_image'] = time.time() - start_time
        return self
        
    def detect_circle(self):
        """使用CircleDetector检测圆"""
        if self.final_binary is None:
            raise ValueError("未完成图像处理")
        
        start_time = time.time()
        
        # 将我们处理后的二值图传给circle_detector
        self.circle_detector.processed_image = self.final_binary
        
        # 使用circle_detector进行检测
        result = self.circle_detector.detect_circle()
        
        # 在原始图像上绘制结果
        self.result_image = self.original_image.copy()
        
        if result:
            center, radius = result
            # 绘制圆和圆心
            cv2.circle(self.result_image, center, radius, (0, 255, 0), 2)
            cv2.circle(self.result_image, center, 2, (0, 0, 255), 3)
            
        self.timing['detect_circle'] = time.time() - start_time
        return result

    def print_timing_stats(self):
        """打印时间统计信息"""
        print("\n=== Performance Statistics ===")
        total_time = sum(self.timing.values())
        
        # 打印每个步骤的时间
        for step, duration in self.timing.items():
            percentage = (duration / total_time) * 100
            print(f"{step:15s}: {duration*1000:6.2f} ms ({percentage:5.1f}%)")
        
        # 打印总时间
        print("-" * 40)
        print(f"{'Total':15s}: {total_time*1000:6.2f} ms (100.0%)")
        print(f"FPS          : {1/total_time:6.2f}")

    def show_process(self):
        """显示处理过程"""
        if self.original_image is None:
            raise ValueError("未加载图像")
            
        # 创建显示窗口
        cv2.namedWindow('Processing Steps', cv2.WINDOW_NORMAL)
        
        # 准备所有处理步骤的图像
        images = []
        titles = []
        
        # 添加原始图像
        images.append(self.original_image)
        titles.append('1. Original Image')
        
        # 添加初始二值图像
        if self.initial_binary is not None:
            initial_binary_display = cv2.cvtColor(self.initial_binary, cv2.COLOR_GRAY2BGR)
            images.append(initial_binary_display)
            titles.append('2. Initial Binary')
        
        # 添加滤波后的二值图像
        if self.binary_image is not None:
            binary_display = cv2.cvtColor(self.binary_image, cv2.COLOR_GRAY2BGR)
            images.append(binary_display)
            titles.append('3. Filtered Binary')
        
        # 添加腐蚀结果
        if self.eroded_image is not None:
            eroded_display = cv2.cvtColor(self.eroded_image, cv2.COLOR_GRAY2BGR)
            images.append(eroded_display)
            titles.append('4. Eroded Image')
        
        # 添加轮廓分析结果
        if self.contour_image is not None:
            images.append(self.contour_image)
            titles.append('5. Contour Analysis')
        
        # 添加过滤后的图像
        if self.filtered_image is not None:
            filtered_display = cv2.cvtColor(self.filtered_image, cv2.COLOR_GRAY2BGR)
            images.append(filtered_display)
            titles.append('6. Filtered Contours')
        
        # 添加最终的二值图像
        if self.final_binary is not None:
            final_display = cv2.cvtColor(self.final_binary, cv2.COLOR_GRAY2BGR)
            images.append(final_display)
            titles.append('7. Final Binary')
        
        # 添加结果图像
        if self.result_image is not None:
            images.append(self.result_image)
            titles.append('8. Detection Result')
        
        # 计算显示布局
        n_images = len(images)
        n_cols = min(4, n_images)  # 每行最多显示4张图
        n_rows = (n_images + n_cols - 1) // n_cols
        
        # 调整图像大小
        height = 300
        processed_images = []
        for img in images:
            aspect_ratio = img.shape[1] / img.shape[0]
            width = int(height * aspect_ratio)
            resized = cv2.resize(img, (width, height))
            processed_images.append(resized)
        
        # 创建画布
        max_width = max(img.shape[1] for img in processed_images)
        canvas = np.zeros((height * n_rows, max_width * n_cols, 3), dtype=np.uint8)
        
        # 放置图像和标题
        for idx, (img, title) in enumerate(zip(processed_images, titles)):
            i, j = idx // n_cols, idx % n_cols
            y_start = i * height
            x_start = j * max_width + (max_width - img.shape[1]) // 2
            
            # 放置图像
            canvas[y_start:y_start+img.shape[0], x_start:x_start+img.shape[1]] = img
            
            # 添加标题
            cv2.putText(canvas, title, (x_start + 10, y_start + 30),
                       cv2.FONT_HERSHEY_SIMPLEX, 0.8, (0, 255, 0), 2)
        
        # 显示结果
        cv2.imshow('Processing Steps', canvas)
        cv2.waitKey(0)
        cv2.destroyAllWindows()

def main():
    # 创建检测器实例
    detector = HoopDetector(
        kernel_size=5,
        solidity_threshold=0.6,
        min_score=0.7
    )
    
    try:
        # 重置时间统计
        detector.reset_timing()
        
        # 加载并处理图像
        detector.load_image("/home/niu/Desktop/wfy/pictures/hoopA.bmp")
        detector.create_binary_image()
        detector.process_image()
        result = detector.detect_circle()
        
        if result:
            center, radius = result
            print(f"检测到篮筐!")
            print(f"圆心坐标: ({center[0]}, {center[1]})")
            print(f"半径: {radius}")
        else:
            print("未检测到篮筐")
        
        # 打印时间统计
        detector.print_timing_stats()
        
        # 显示处理过程
        detector.show_process()
        
    except Exception as e:
        print(f"错误: {str(e)}")

if __name__ == "__main__":
    main() 

File: hoop_detector.cpp
#include "hoop_detector.hpp"
#include <iostream>
#include <random>
#include <vector>
#include <cmath>

void HoopDetector::logIfNeeded(const std::string& message) {
    auto current_time = std::chrono::steady_clock::now();
    double elapsed = std::chrono::duration<double>(current_time - last_log_time_).count();
    if (elapsed >= log_interval_) {
        std::cout << message << std::endl;
        last_log_time_ = current_time;
    }
}

HoopDetector::HoopDetector(int kernel_size, double solidity_threshold, double min_score)
    : circle_detector_(6, min_score)
    , kernel_size_(kernel_size)
    , solidity_threshold_(solidity_threshold)
    , last_save_time_(std::chrono::steady_clock::now())
    , last_log_time_(std::chrono::steady_clock::now()) {
    std::cout << "[HoopDetector] 初始化参数: kernel_size=" << kernel_size 
              << ", solidity_threshold=" << solidity_threshold 
              << ", min_score=" << min_score << std::endl;
    kernel_ = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(kernel_size_, kernel_size_));
}

void HoopDetector::resetTiming() {
    timing_.clear();
}

HoopDetector& HoopDetector::loadImage(const std::string& image_path) {
    auto start = std::chrono::high_resolution_clock::now();
    
    original_image_ = cv::imread(image_path);
    if (original_image_.empty()) {
        throw std::runtime_error("Failed to load image from " + image_path);
    }
    circle_detector_.loadImage(original_image_);
    
    auto end = std::chrono::high_resolution_clock::now();
    timing_["load_image"] = std::chrono::duration<double>(end - start).count();
    
    return *this;
}

HoopDetector& HoopDetector::loadImage(const cv::Mat& image) {
    auto start = std::chrono::high_resolution_clock::now();
    
    if (image.empty()) {
        throw std::runtime_error("[HoopDetector] 输入图像为空");
    }
    
    // 确保图像是BGR格式
    if (image.type() != CV_8UC3) {
        if (image.channels() == 3) {
            cv::Mat temp;
            image.convertTo(temp, CV_8UC3);
            original_image_ = temp.clone();
        } else {
            throw std::runtime_error("[HoopDetector] 不支持的图像格式");
        }
    } else {
        original_image_ = image.clone();
    }
    
    circle_detector_.loadImage(original_image_);
    
    auto end = std::chrono::high_resolution_clock::now();
    timing_["load_image"] = std::chrono::duration<double>(end - start).count();
    
    return *this;
}

HoopDetector& HoopDetector::createBinaryImage() {
    if (original_image_.empty()) {
        throw std::runtime_error("[HoopDetector] 未加载图像");
    }
    
    auto start = std::chrono::high_resolution_clock::now();
    
    try {
        logIfNeeded("[HoopDetector] 开始图像处理...");
        
        // 转换到HSV颜色空间
        cv::Mat hsv_frame;
        cv::cvtColor(original_image_, hsv_frame, cv::COLOR_BGR2HSV);
        
        // 创建橙色掩码
        cv::Scalar lower_bound(0, 125, 72);
        cv::Scalar upper_bound(13, 255, 153);
        cv::inRange(hsv_frame, lower_bound, upper_bound, initial_binary_);
        
        if (initial_binary_.empty()) {
            throw std::runtime_error("[HoopDetector] 二值图创建失败");
        }
        
        // 清理掩码
        // cv::Mat element = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(5, 5));
        // cv::morphologyEx(initial_binary_, initial_binary_, cv::MORPH_CLOSE, element);
        cv::medianBlur(initial_binary_, binary_image_, 23);
        
    } catch (const cv::Exception& e) {
        std::cerr << "[HoopDetector] OpenCV错误: " << e.what() << std::endl;
        throw;
    }
    
    auto end = std::chrono::high_resolution_clock::now();
    timing_["create_binary"] = std::chrono::duration<double>(end - start).count();
    
    return *this;
}

HoopDetector& HoopDetector::processImage() {
    if (binary_image_.empty()) {
        throw std::runtime_error("[HoopDetector] Binary image not created");
    }
    
    auto start = std::chrono::high_resolution_clock::now();
    
    try {
        std::cout << "[HoopDetector] 开始处理图像..." << std::endl;
        
        // 1. 直接使用binary_image_作为最终的二值图
        final_binary_ = binary_image_.clone();
        
        // 2. 保存中间结果供显示使用
        filtered_image_ = binary_image_.clone();  // 保存滤波后的二值图
        
        std::cout << "[HoopDetector] 图像处理完成" << std::endl;
        
    } catch (const cv::Exception& e) {
        std::cerr << "[HoopDetector] OpenCV错误: " << e.what() << std::endl;
        throw;
    } catch (const std::exception& e) {
        std::cerr << "[HoopDetector] 标准错误: " << e.what() << std::endl;
        throw;
    } catch (...) {
        std::cerr << "[HoopDetector] 未知错误!" << std::endl;
        throw;
    }
    
    auto end = std::chrono::high_resolution_clock::now();
    timing_["process_image"] = std::chrono::duration<double>(end - start).count();
    
    return *this;
}

std::pair<cv::Point, int> HoopDetector::detectCircle() {
    if (final_binary_.empty()) {
        throw std::runtime_error("[HoopDetector] 图像未处理");
    }
    
    auto start = std::chrono::high_resolution_clock::now();
    
    try {
        logIfNeeded("[HoopDetector] 开始圆检测...");
        
        // 提取所有轮廓点到一个向量中
        std::vector<cv::Point> all_contour_points;
        std::vector<std::vector<cv::Point>> contours;
        std::vector<cv::Vec4i> hierarchy;
        cv::findContours(final_binary_, contours, hierarchy, 
                        cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
        
        // 将所有轮廓的点合并到一个向量中
        for(const auto& contour : contours) {
            all_contour_points.insert(all_contour_points.end(), contour.begin(), contour.end());
        }
        
        logIfNeeded("[HoopDetector] 总共提取了 " + std::to_string(all_contour_points.size()) + " 个轮廓点");
        
        
        // 在原始图像和二值图上绘制结果
        result_image_ = original_image_.clone();
        cv::Mat binary_result;
        cv::cvtColor(final_binary_, binary_result, cv::COLOR_GRAY2BGR);
        
        // // 找到最大的轮廓
        // int max_contour_idx = -1;
        // int max_contour_size = 0;
        // for (size_t i = 0; i < contours.size(); i++) {
        //     int size = contours[i].size();
        //     if (size > max_contour_size) {
        //         max_contour_size = size;
        //         max_contour_idx = i;
        //     }
        // }
        
        // if (max_contour_idx >= 0) {
            // 使用RANSAC拟合圆

            auto result = fitCircleRANSAC(all_contour_points);
            auto center = result.first;
            auto radius = result.second;
            
            if (radius > 0) {
                logIfNeeded(
                    "[HoopDetector] 检测结果: 中心=(" + std::to_string(center.x) + "," + 
                    std::to_string(center.y) + ") 半径=" + std::to_string(radius)
                );
                
                // 在原始图像上绘制加粗的圆和更大的圆心
                cv::circle(result_image_, center, radius, cv::Scalar(0, 255, 0), 4);
                cv::circle(result_image_, center, 5, cv::Scalar(0, 0, 255), -1);
                
                // 在二值图上也绘制加粗的圆和更大的圆心
                cv::circle(binary_result, center, radius, cv::Scalar(0, 255, 0), 4);
                cv::circle(binary_result, center, 5, cv::Scalar(0, 0, 255), -1);
                
                // 更新final_binary_为带有拟合结果的图像
                cv::cvtColor(binary_result, final_binary_, cv::COLOR_BGR2GRAY);
                
                auto end = std::chrono::high_resolution_clock::now();
                timing_["detect_circle"] = std::chrono::duration<double>(end - start).count();
                
                // 计算并输出性能统计
                double total_time = 0;
                for (const auto& [_, duration] : timing_) {
                    total_time += duration;
                }
                
                logIfNeeded(
                    "[HoopDetector] 性能统计: 总耗时=" + std::to_string(total_time * 1000) + 
                    "ms, FPS=" + std::to_string(1.0 / total_time)
                );
                
                return std::make_pair(center, radius);
            }
        
        
        logIfNeeded("[HoopDetector] 未检测到圆");
        return std::make_pair(cv::Point(0, 0), 0);
        
    } catch (const cv::Exception& e) {
        std::cerr << "[HoopDetector] OpenCV错误: " << e.what() << std::endl;
        throw;
    } catch (const std::exception& e) {
        std::cerr << "[HoopDetector] 标准错误: " << e.what() << std::endl;
        throw;
    } catch (...) {
        std::cerr << "[HoopDetector] 未知错误!" << std::endl;
        throw;
    }
}

void HoopDetector::printTimingStats() const {
    double total_time = 0;
    for (const auto& [_, duration] : timing_) {
        total_time += duration;
    }
    
    std::cout << "\n=== 性能统计 ===" << std::endl;
    for (const auto& [step, duration] : timing_) {
        double percentage = (duration / total_time) * 100;
        printf("%-15s: %6.2f ms (%5.1f%%)\n", 
               step.c_str(), duration * 1000, percentage);
    }
    
    std::cout << std::string(40, '-') << std::endl;
    printf("%-15s: %6.2f ms\n", "总耗时", total_time * 1000);
    printf("FPS          : %6.2f\n", 1.0 / total_time);
}

cv::Mat HoopDetector::showProcess() const {
    if (original_image_.empty()) {
        throw std::runtime_error("No image loaded");
    }
    
    // 准备所有处理步骤的图像
    std::vector<cv::Mat> images;
    std::vector<std::string> titles;
    
    // 添加原始图像
    images.push_back(original_image_);
    titles.push_back("1. Original Image");
    
    // 添加初始二值图像
    if (!initial_binary_.empty()) {
        cv::Mat initial_binary_display;
        cv::cvtColor(initial_binary_, initial_binary_display, cv::COLOR_GRAY2BGR);
        images.push_back(initial_binary_display);
        titles.push_back("2. Initial Binary");
    }
    
    // 添加滤波后的二值图像
    if (!binary_image_.empty()) {
        cv::Mat binary_display;
        cv::cvtColor(binary_image_, binary_display, cv::COLOR_GRAY2BGR);
        images.push_back(binary_display);
        titles.push_back("3. Filtered Binary");
    }
    
    // 添加最终的二值图像
    if (!final_binary_.empty()) {
        cv::Mat final_display;
        cv::cvtColor(final_binary_, final_display, cv::COLOR_GRAY2BGR);
        images.push_back(final_display);
        titles.push_back("4. Final Binary");
    }
    
    // 添加结果图像
    if (!result_image_.empty()) {
        images.push_back(result_image_);
        titles.push_back("5. Detection Result");
    }
    
    // 计算显示布局
    int n_images = images.size();
    int n_cols = std::min(3, n_images);  // 每行最多显示3张图
    int n_rows = (n_images + n_cols - 1) / n_cols;
    
    // 调整图像大小
    const int height = 300;
    std::vector<cv::Mat> processed_images;
    int max_width = 0;
    
    for (const auto& img : images) {
        double aspect_ratio = static_cast<double>(img.cols) / img.rows;
        int width = static_cast<int>(height * aspect_ratio);
        cv::Mat resized;
        cv::resize(img, resized, cv::Size(width, height));
        processed_images.push_back(resized);
        max_width = std::max(max_width, width);
    }
    
    // 创建画布
    cv::Mat canvas = cv::Mat::zeros(height * n_rows, max_width * n_cols, CV_8UC3);
    
    // 放置图像和标题
    for (int idx = 0; idx < n_images; ++idx) {
        int i = idx / n_cols;
        int j = idx % n_cols;
        int y_start = i * height;
        int x_start = j * max_width + (max_width - processed_images[idx].cols) / 2;
        
        // 放置图像
        processed_images[idx].copyTo(
            canvas(cv::Rect(x_start, y_start, 
                          processed_images[idx].cols, 
                          processed_images[idx].rows)));
        
        // 添加标题
        cv::putText(canvas, titles[idx], 
                   cv::Point(x_start + 10, y_start + 30),
                   cv::FONT_HERSHEY_SIMPLEX, 0.8, 
                   cv::Scalar(0, 255, 0), 2);
    }
    
    return canvas;
}

std::pair<cv::Point, int> HoopDetector::fitCircleRANSAC(
    const std::vector<cv::Point>& points, 
    int n_iterations, 
    float threshold_dist) {
    
    if (points.size() < 3) {
        return std::make_pair(cv::Point(0, 0), 0);
    }

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, points.size() - 1);

    cv::Point best_center(0, 0);
    int best_radius = 0;
    int max_inliers = 0;

    for (int iter = 0; iter < n_iterations; ++iter) {
        // 随机选择3个点
        int idx1 = dis(gen);
        int idx2 = dis(gen);
        int idx3 = dis(gen);

        // 确保选择的点不重复
        if (idx1 == idx2 || idx2 == idx3 || idx1 == idx3) {
            continue;
        }

        cv::Point2f p1 = points[idx1];
        cv::Point2f p2 = points[idx2];
        cv::Point2f p3 = points[idx3];

        // 计算三点确定的圆
        float x1 = p1.x, y1 = p1.y;
        float x2 = p2.x, y2 = p2.y;
        float x3 = p3.x, y3 = p3.y;

        float a = x1 * (y2 - y3) - y1 * (x2 - x3) + x2 * y3 - x3 * y2;
        float b = (x1 * x1 + y1 * y1) * (y3 - y2) + 
                 (x2 * x2 + y2 * y2) * (y1 - y3) + 
                 (x3 * x3 + y3 * y3) * (y2 - y1);
        float c = (x1 * x1 + y1 * y1) * (x2 - x3) + 
                 (x2 * x2 + y2 * y2) * (x3 - x1) + 
                 (x3 * x3 + y3 * y3) * (x1 - x2);

        // 检查是否有效的圆
        if (std::abs(a) < 1e-7) {
            continue;
        }

        float center_x = -b / (2 * a);
        float center_y = -c / (2 * a);
        cv::Point2f center(center_x, center_y);

        // 计算半径
        float radius = std::sqrt(std::pow(x1 - center_x, 2) + std::pow(y1 - center_y, 2));

        // 计算内点数量
        int inliers = 0;
        for (const auto& point : points) {
            float dist = std::abs(std::sqrt(std::pow(point.x - center_x, 2) + 
                                          std::pow(point.y - center_y, 2)) - radius);
            if (dist < threshold_dist) {
                inliers++;
            }
        }

        // 更新最佳结果
        if (inliers > max_inliers && inliers > 150) {
            max_inliers = inliers;
            best_center = cv::Point(std::round(center_x), std::round(center_y));
            best_radius = std::round(radius);
        }
    }

    return std::make_pair(best_center, best_radius);
} 